src/katgpucbf/xbgpu/engine.py

```diff
@@ -117,6 +127,417 @@ def reset(self, timestamp: int = 0) -> None:
         self.batches = 0
 
 
+class Pipeline:
+    """Base Pipeline class to build on.
+
+    Parameters
+    ----------
+    name
+        Name of Pipeline.
+    engine
+        The owning engine.
+    context
+        CUDA context for device work.
+    """
+
+    # NOTE: n_rx_items and n_tx_items dictate the number of GPU buffers.
+    # Setting these values too high results in too much GPU memory being
+    # consumed. The quantity of each needs to be sufficient so as not to starve
+    # the different processing loops. The low single digits is suitable.
+    # These values are not configurable as they have been acceptable for
+    # most tests cases up until now. If the pipeline starts bottlenecking,
+    # then maybe look at increasing these values.
+    n_rx_items = DEFAULT_N_RX_ITEMS
+    n_tx_items = DEFAULT_N_TX_ITEMS  # TODO: Will likely need to be == n_rx_items for BPipeline
+
+    def __init__(self, name: str, engine: "XBEngine", context: AbstractContext) -> None:
+        self.engine = engine
+        self.name = name
+
+        self._proc_command_queue = context.create_command_queue()
+        self._download_command_queue = context.create_command_queue()
+
+        # These queues are extended in the monitor class, allowing for the
+        # monitor to track the number of items on each queue.
+        # - The _rx_item_queue receives items from :meth:`.XBEngine._receiver_loop`
+        #   to be used by :meth:`_gpu_proc_loop`.
+        # - The _tx_item_queue receives items from :meth:`gpu_proc_loop` to be
+        #   used by :meth:`sender_loop`.
+        # Once the destination function is finished with an item, it will pass
+        # it back to the corresponding _(rx/tx)_free_item_queue to ensure that
+        # all allocated buffers are in continuous circulation.
+        # TODO: BPipeline may/may not adopt this 1:1 {rx, tx}_item queue approach
+        self._rx_item_queue: asyncio.Queue[RxQueueItem | None] = engine.monitor.make_queue(
+            f"{name}.rx_item_queue", self.n_rx_items
+        )
+        self._tx_item_queue: asyncio.Queue[TxQueueItem | None] = engine.monitor.make_queue(
+            f"{name}.tx_item_queue", self.n_tx_items
+        )
+        self._tx_free_item_queue: asyncio.Queue[TxQueueItem] = engine.monitor.make_queue(
+            f"{name}.tx_free_item_queue", self.n_tx_items
+        )
+
+    def add_rx_item(self, item: RxQueueItem) -> None:
+        """Append a newly-received :class:`RxQueueItem` to the :attr:`_rx_item_queue`."""
+        self._rx_item_queue.put_nowait(item)
+
+    def shutdown(self) -> None:
+        """Start a graceful shutdown after the final call to :meth:`add_rx_item`."""
+        self._rx_item_queue.put_nowait(None)
+
+    @abstractmethod
+    async def gpu_proc_loop(self) -> None:
+        """Perform all GPU processing of received data in a continuous loop.
+
+        This function should do the following:
```

Comment:
```suggestion
        This method does the following:
```

I assume you've written "should do" because this is the abstract declaration, but the same docstring gets inherited by the implementation, and it's confusing when consulting the XPipeline documentation. Same applies to the other abstract methods in this class.